honeybee_3dm/togeometry.py

```diff
@@ -1,34 +1,309 @@
+                start_pt = to_point3d(brep.Edges[i].PointAtStart)
+                end_pt = to_point3d(brep.Edges[i].PointAtEnd)
+                line = LineSegment3D.from_end_points(start_pt, end_pt)
+                lines.append(line)
+
+            # Create Ladybug Polylines from the lines
+            polylines = Polyline3D.join_segments(lines, tolerance)
+            # Sort the Ladybug Polylines based on length
+            # The longest polyline belongs to the boundary of the face
+            # The rest of the polylines belong to the holes in the face
+            sorted_polylines = sorted(polylines, key=lambda polyline: polyline.length)
+            # The first polyline in the list shall always be the polyline for
+            # the boundary
+            sorted_polylines.reverse()
+
+            # Getting all vertices of the face
+            mesh = brep.Faces[0].GetMesh(rhino3dm.MeshType.Any)
+            mesh_pts = [to_point3d(mesh.Vertices[i])
+                        for i in range(len(mesh.Vertices))]
+
+            # In the list of the Polylines, if there's only one polyline then
+            # the face has no holes
+            if len(sorted_polylines) == 1:
+                if len(mesh_pts) == 4 or len(mesh_pts) == 3:
+                    faces.extend(mesh_to_face3d(mesh))
+                else:
+                    boundary_pts = remove_dup_vertices(
+                        sorted_polylines[0].vertices, tolerance)
+                    lb_face = Face3D(boundary=boundary_pts)
+                    faces.append(lb_face)
+
+            # In the list of the Polylines, if there's more than one polyline then
+            # the face has hole / holes
+            elif len(sorted_polylines) > 1:
+                boundary_pts = remove_dup_vertices(
+                    sorted_polylines[0].vertices, tolerance)
+                hole_pts = [remove_dup_vertices(polyline.vertices, tolerance)
+                            for polyline in sorted_polylines[1:]]
+                # Merging lists of hole_pts
+                total_hole_pts = [
+                    pts for pts_lst in hole_pts for pts in pts_lst]
+                hole_pts_on_boundary = [
+                    pts for pts in total_hole_pts if pts in boundary_pts]
+                # * Check 02 - If any of the hole is touching the boundary of the face
+                # then mesh it
+                if len(hole_pts_on_boundary) > 0:
+                    warnings.warn(
+                        "The surface has holes that touch the boundary. \
+                            It will be meshed")
+                    faces.extend(brep_to_face3d(brep))
+                else:
+                    lb_face = Face3D(boundary=boundary_pts, holes=hole_pts)
+                    faces.append(lb_face)
+    # If any of the above fails, mesh it
+    except Exception as e:
+        print(str(e) + "Face creation failed, this face will be meshed")
+        faces.extend(brep_to_face3d(brep))
```

Comment:
This will be confusing. We should at least separate them in outputs and let the user handle the objects with different types.

Reply:
@mostaphaRoudsari,  In this method we have two checks [here](https://github.com/devngc/honeybee-3dm/blob/master/honeybee_3dm/togeometry.py#L159) and [here](https://github.com/devngc/honeybee-3dm/blob/master/honeybee_3dm/togeometry.py#L216). I have kept this exception to catch any other scenario for which we have not created a check yet. I would remove this message and let the user see the error message instead.

Do you also want me to report the object to the user? I can print the object ID and then a user can use `SelID` in rhino to quickly select the object.

Reply:
I follow your logic and I see what are you trying to do but it will not address my initial conerns. This is more of a "code smell" than a bug. The structure of the code is not clean enough. If you are expecting an error it should be checked where the error happens.

```python
try:
    # small code here
expect ValueError:
   # handle the error
else:
   # no error - do what needs to be done
```

You can have multiple of these instead of one. In the current code, so many things can go wrong and a single try/expect is not necessarily the best practice. I'm not familiar enough with the SDK to comment on details. That's why I asked for @chriswmackey's help here.

Reply:
@mostaphaRoudsari ,
I changed this and made the exception handling specific.